src/main/java/com/snowplowanalytics/snowplow/tracker/emitter/BatchEmitter.java

```diff
@@ -150,61 +165,67 @@ public int getBufferSize() {
-     *
-     * @return the new Runnable object
-     */
-    private Runnable getCheckForEventsToSendRunnable() {
-        return () -> {
-            while (!isClosing) {
-                if (eventStore.getSize() >= bufferSize) {
-                    drainEventsAndSend(getBufferSize());
-                }
-            }
-        };
-    }
-
-    private void drainEventsAndSend(int numberOfEvents) {
-        List<TrackerPayload> payloads = eventStore.removeEvents(numberOfEvents);
-        execute(getPostRequestRunnable(payloads));
+    public long getRetryDelay() {
+        return this.retryDelay.get();
     }
 
     /**
      * Returns a Runnable POST Request operation
      *
-     * @param buffer the event buffer to be sent
+     * @param numberOfEvents the number of events to be sent in the request
      * @return the new Runnable object
      */
-    private Runnable getPostRequestRunnable(final List<TrackerPayload> buffer) {
+    private Runnable getPostRequestRunnable(int numberOfEvents) {
         return () -> {
-            if (buffer.size() == 0) {
-                return;
-            }
+            BatchPayload batchedEvents = null;
+            try {
+                batchedEvents = this.eventStore.getEventBatch(numberOfEvents);
+                List<TrackerPayload> eventsInRequest = batchedEvents.getPayloads();
 
-            final SelfDescribingJson post = getFinalPost(buffer);
-            final int code = httpClientAdapter.post(post);
+                if (eventsInRequest.size() == 0) {
+                    return;
+                }
 
-            // Process results
-            if (!isSuccessfulSend(code)) {
-                LOGGER.error("BatchEmitter failed to send {} events: code: {}", buffer.size(), code);
-            } else {
-                LOGGER.debug("BatchEmitter successfully sent {} events: code: {}", buffer.size(), code);
+                final SelfDescribingJson post = getFinalPost(eventsInRequest);
+                final int code = httpClientAdapter.post(post);
+
+                // Process results
+                if (isSuccessfulSend(code)) {
+                    LOGGER.debug("BatchEmitter successfully sent {} events: code: {}", eventsInRequest.size(), code);
+                    this.retryDelay.set(0L);
+                    eventStore.cleanupAfterSendingAttempt(true, batchedEvents.getBatchId());
+                } else {
+                    LOGGER.error("BatchEmitter failed to send {} events: code: {}", eventsInRequest.size(), code);
+                    eventStore.cleanupAfterSendingAttempt(false, batchedEvents.getBatchId());
+
+                    // exponentially increase retry backoff time after the first failure
+                    long currentDelay = this.retryDelay.get();
+                    if (currentDelay == 0) {
+                        this.retryDelay.compareAndSet(0, 50L);
+                    } else {
+                        this.retryDelay.set(currentDelay * 2);
+                    }
```

Comment:
I think doing that you don't make use of the atomicity of `retryDelay`. I would suggest something link this:
```suggestion
                    if (!retryDelay.compareAndSet(0, 50L)) {
                        retryDelay.updateAndGet(currentDelay -> currentDelay * 2);
                    }
```